### PR TITLE
#215 | Add soft-deletion support for Varadhi Resources

### PR DESCRIPTION
**Pull request by devanshu-garg-IN**
# Description [#215]

This PR introduces support for soft and hard deletion for Varadhi resources, enabling flexible lifecycle management for topics and subscriptions. It includes updates to the service layer, deletion handler, restoration process, and unit tests to ensure a robust and maintainable implementation.

## Key Changes

Here are the release notes:

### New Features

- Added support for soft and hard deletion of topics and subscriptions.
- Introduced topic and subscription status management (active/inactive).
- Added topic and subscription restoration capabilities.
- Enhanced resource management with new deletion type handling.
- Introduced new methods for managing subscription states and validations.
- New `LifecycleStatus` class to encapsulate lifecycle states and action codes.
- Added `ResourceActionRequest` record for action requests on resources.

### Improvements

- Improved validation and error handling across services.
- Refined logging and error messages for clarity.
- Enhanced method naming and code structure for better readability.
- Added more comprehensive test coverage for various scenarios.
- Updated `SubscriptionResource` and `TopicResource` to include actor codes.

### Bug Fixes

- Improved handling of resource states and transitions.
- Fixed potential edge cases in resource management.

### Technical Enhancements

- Updated constant naming conventions for consistency.
- Refactored service and handler classes for better maintainability.
- Added more robust input validation mechanisms.

# Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [x] This change requires a documentation update

# Testing and Validation

- Added comprehensive unit tests to verify the correctness of the implementation.
- Included edge cases to ensure robust behavior, such as:

  - Proper handling of inactive resources.
  - Managing failures in the meta store.

- Refactored test code for improved maintainability and readability.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

## Summary by CodeRabbit

- **New Features**
  - Introduced robust lifecycle management for topics and subscriptions, including support for soft/hard delete and restore operations.
  - Added support for HTTP PATCH requests within the API.
  - Added new `ResourceActionRequest` record to encapsulate action requests with actor codes and messages.

- **Enhancements**
  - Improved validation and error messaging to provide clearer, more user-friendly responses.
  - Streamlined resource naming and state management for more consistent operations.
  - Expanded testing and documentation to ensure increased stability and reliability.
  - Updated subscription and topic resource handling to include lifecycle status action codes.
  - Enhanced `VaradhiSubscriptionFactory` and `VaradhiTopicFactory` with improved methods for shard management and topic creation.
  - Improved encapsulation and clarity in `SubscriptionResource` and `TopicResource` classes by adding new fields and modifying constructors.

### CODE REVIEW

File: entities/src/main/java/com/flipkart/varadhi/entities/LifecycleStatus.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+        UPDATING("Update in progress."),
+        UPDATED("Successfully updated."),
+        UPDATE_FAILED("Update failed."),
+
+        DELETING("Deletion in progress."),
+        DELETED("Successfully deleted."),
+        DELETE_FAILED("Deletion failed."),
+
+        ACTIVE("Currently active."),
```
Comment (gauravAshok) [inline entities/src/main/java/com/flipkart/varadhi/entities/LifecycleStatus.java:87]: I understand, create flow & delete flow & inactive state.
But how does states for update flow help?
How is active state different from created state? How is DELETED even a state? Where will it be used?

File: entities/src/main/java/com/flipkart/varadhi/entities/LifecycleStatus.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+    /**
+     * The message associated with the current state.
+     */
+    private String message;
+
+    /**
+     * The action code indicating the reason for the current state.
+     */
+    private ActionCode actionCode;
```
Comment (gauravAshok) [inline entities/src/main/java/com/flipkart/varadhi/entities/LifecycleStatus.java:29]:
```suggestion
    private ActorCode actorCode;
```
?

The actor of the "action" :)

File: entities/src/main/java/com/flipkart/varadhi/entities/ResourceActionRequest.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+import jakarta.validation.constraints.NotNull;
+
+/**
+ * Represents a request for an action on a resource, containing an action code and an optional message.
+ *
+ * @param actionCode the action code indicating the action to be performed, must not be null
+ * @param message    an optional message associated with the action
+ */
+public record ResourceActionRequest(@NotNull LifecycleStatus.ActionCode actionCode, String message)
```
Comment (gauravAshok) [inline entities/src/main/java/com/flipkart/varadhi/entities/ResourceActionRequest.java:11]: should this be part of the entities module? can this be moved to server/web module?

File: entities/src/test/java/com/flipkart/varadhi/entities/auth/ResourceActionTest.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+                () -> assertEquals("varadhi.project.delete", ResourceAction.PROJECT_DELETE.toString()),
+                () -> assertEquals("varadhi.topic.get", ResourceAction.TOPIC_GET.toString()),
+                () -> assertEquals("varadhi.project.list", ResourceAction.SUBSCRIPTION_LIST.toString()),
+                () -> assertEquals("varadhi.iam_policy.set", ResourceAction.IAM_POLICY_SET.toString())
+        );
+    }
+
+    @Test
+    void actionName_ReturnsCorrectActionName() {
+        assertAll(
```
Comment (gauravAshok) [inline entities/src/test/java/com/flipkart/varadhi/entities/auth/ResourceActionTest.java:24]: are these tests of any use :)

File: entities/src/main/java/com/flipkart/varadhi/entities/LifecycleStatus.java (hunk 1 above)
Comment (devanshu-garg-IN) [reply entities/src/main/java/com/flipkart/varadhi/entities/LifecycleStatus.java:87]: I have currently included all possible states, including `active` and `inactive`, for completeness. Completely aligned that **"deleted" is not a valid state**, and that **"created" and "active" are essentially the same**.

This PR primarily focuses on **deletion support**, and in the **next PR**, where I'll be implementing **idempotency across handlers**, I'll refine these states and use them appropriately at the REST layer.

For **update operations**, the state management will come into play once **idempotency is introduced** in the next iteration. However, the **"updated"** state itself won’t be needed—only **"updating"** and **"update failed"** will be used, since **"updated" is functionally equivalent to "created"**.

For now, these states help in **decoupling lifecycle status from individual resource implementations**.

File: entities/src/main/java/com/flipkart/varadhi/entities/LifecycleStatus.java (hunk 2 above)
Comment (devanshu-garg-IN) [reply entities/src/main/java/com/flipkart/varadhi/entities/LifecycleStatus.java:29]: Agreed.

File: entities/src/main/java/com/flipkart/varadhi/entities/ResourceActionRequest.java (hunk 3 above)
Comment (devanshu-garg-IN) [reply entities/src/main/java/com/flipkart/varadhi/entities/ResourceActionRequest.java:11]: Yes, it can be moved to Server Entities. I'll do it.

File: entities/src/test/java/com/flipkart/varadhi/entities/auth/ResourceActionTest.java (hunk 4 above)
Comment (devanshu-garg-IN) [reply entities/src/test/java/com/flipkart/varadhi/entities/auth/ResourceActionTest.java:24]: Agreed! While the suggested approach works, `JaCoCo`, `Cobertura`, `Emma`, **or similar tools flag it as uncovered code**, meaning it doesn't contribute to test coverage. That’s why I’ve explicitly added all possible test cases to **maximize coverage**.

Comment (gauravAshok) [reply entities/src/test/java/com/flipkart/varadhi/entities/auth/ResourceActionTest.java:24]: na. that shouldn;t happen. they literally work by tracing the execution, as long as execution happened, the coverage should reflect it.

File: entities/src/main/java/com/flipkart/varadhi/entities/ResourceDeletionType.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
+package com.flipkart.varadhi.entities;
+
+/**
+ * Enum representing the types of resource deletion.
+ */
+public enum ResourceDeletionType {
+    SOFT_DELETE,
+    HARD_DELETE,
+    DEFAULT;
```
Comment (gauravAshok) [inline entities/src/main/java/com/flipkart/varadhi/entities/ResourceDeletionType.java:9]: dont think DEFAULT is needed here.

File: entities/src/main/java/com/flipkart/varadhi/entities/ResourceDeletionType.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    /**
+     * Converts a string value to a ResourceDeletionType.
+     *
+     * @param value the string value to convert
+     *
+     * @return the corresponding ResourceDeletionType, or DEFAULT if the value is null, blank, or unrecognized
+     */
+    public static ResourceDeletionType fromValue(String value) {
+        if (value == null || value.isBlank()) {
+            return DEFAULT;
```
Comment (gauravAshok) [inline entities/src/main/java/com/flipkart/varadhi/entities/ResourceDeletionType.java:20]: i think you can choose among the SOFT / HARD ehre.

File: entities/src/main/java/com/flipkart/varadhi/entities/LifecycleStatus.java (hunk 2 above)
Comment (devanshu-garg-IN) [reply entities/src/main/java/com/flipkart/varadhi/entities/LifecycleStatus.java]: Done

File: entities/src/main/java/com/flipkart/varadhi/entities/ResourceActionRequest.java (hunk 3 above)
Comment (devanshu-garg-IN) [reply entities/src/main/java/com/flipkart/varadhi/entities/ResourceActionRequest.java]: Done

File: entities/src/test/java/com/flipkart/varadhi/entities/auth/ResourceActionTest.java (hunk 4 above)
Comment (devanshu-garg-IN) [reply entities/src/test/java/com/flipkart/varadhi/entities/auth/ResourceActionTest.java:24]: Ahh okay!

File: entities/src/main/java/com/flipkart/varadhi/entities/ResourceDeletionType.java (hunk 5 above)
Comment (devanshu-garg-IN) [reply entities/src/main/java/com/flipkart/varadhi/entities/ResourceDeletionType.java]: Agreed! Done.

File: entities/src/main/java/com/flipkart/varadhi/entities/ResourceDeletionType.java (hunk 6 above)
Comment (devanshu-garg-IN) [reply entities/src/main/java/com/flipkart/varadhi/entities/ResourceDeletionType.java]: Done

File: entities/src/main/java/com/flipkart/varadhi/entities/LifecycleStatus.java (hunk 1 above)
Comment (gauravAshok) [reply entities/src/main/java/com/flipkart/varadhi/entities/LifecycleStatus.java:87]: For now, I will suggest to start minimally, since it is easier to add stuff then rework stuff. We should add stuff that is only required right away.

It also helps with review. Can't really review what you have in mind for future, since my ideas of future may vary from yours :)

Lets only add stuff that is required for this PR specifically, skip everything else. The design should be such, that allows for incremental changes. and those changes will come in future separately.

Comment (devanshu-garg-IN) [reply entities/src/main/java/com/flipkart/varadhi/entities/LifecycleStatus.java:82]: Done.

File: entities/src/main/java/com/flipkart/varadhi/entities/VaradhiSubscription.java [hunk 7]
```diff
@@ -34,28 +56,39 @@ private VaradhiSubscription(
             RetryPolicy retryPolicy,
             ConsumptionPolicy consumptionPolicy,
             SubscriptionShards shards,
-            Status status,
+            LifecycleStatus status,
```
Comment (gauravAshok) [inline entities/src/main/java/com/flipkart/varadhi/entities/VaradhiSubscription.java:59]: entities with lifecycle can probably become a base class. so Topic & Subscription can extend that. and methods to change the status can move to that base class.

File: server/build.gradle [hunk 8]
```diff
@@ -101,3 +101,8 @@ run {
             "-Dlog4j.configurationFile=$buildDir/resources/main/log4j2.xml"
     ]
 }
+
+testE2E {
+    useJUnitPlatform()
+    jvmArgs '--add-opens', 'java.base/java.net=ALL-UNNAMED'
```
Comment (gauravAshok) [inline server/build.gradle:107]: add-opens shouldn't be required.

File: entities/src/main/java/com/flipkart/varadhi/entities/VaradhiSubscription.java (hunk 7 above)
Comment (devanshu-garg-IN) [reply entities/src/main/java/com/flipkart/varadhi/entities/VaradhiSubscription.java:59]: Agreed! I considered that too, but the states didn't fully align equally for both resources, so I didn’t pick it up initially. That said, it’s definitely being addressed as part of #220.

File: server/build.gradle (hunk 8 above)
Comment (devanshu-garg-IN) [reply server/build.gradle:107]: Yeah, ideally it shouldn't be required. But I was facing issues with PATCH requests, and the error `java.lang.reflect.InaccessibleObjectException` indicated that the PATCH method was restricted due to Java module access controls.

File: server/src/main/java/com/flipkart/varadhi/services/SubscriptionService.java [hunk 9]
```diff
@@ -49,89 +110,270 @@ public VaradhiSubscription createSubscription(
...
-        });
+    /**
+     * Checks if a subscription is active or well-provisioned.
+     *
+     * @param subscription the subscription to check
+     *
+     * @return true if the subscription is active or well-provisioned, false otherwise
+     */
+    private boolean isActiveOrWellProvisioned(VaradhiSubscription subscription) {
+        return subscription.isActive() || subscription.isWellProvisioned();
```
Comment (gauravAshok) [inline server/src/main/java/com/flipkart/varadhi/services/SubscriptionService.java:258]: the fact that we have to rely on 2 states, means there is still redundancy.

A good subscription should come from just 1 state. I would have expected CREATED means good.

File: server/src/main/java/com/flipkart/varadhi/services/SubscriptionService.java [hunk 10]
```diff
@@ -1,46 +1,107 @@
...
-        return metaStore.getSubscriptionNames(projectName);
+    /**
+     * Retrieves the list of subscription names for a given project.
+     *
+     * @param projectName the name of the project
+     * @param includeInactive flag to include inactive or soft-deleted subscriptions
+     *
+     * @return the list of subscription names
+     */
+    public List<String> getSubscriptionList(String projectName, boolean includeInactive) {
```
Comment (gauravAshok) [inline server/src/main/java/com/flipkart/varadhi/services/SubscriptionService.java:57]: this list will include under creation subscription also. 

Bye default the list should onlt include "CREATED". the. flag should allow to include everything else as well.

File: server/src/main/java/com/flipkart/varadhi/services/SubscriptionService.java [hunk 11]
```diff
@@ -1,46 +1,107 @@
...
+    }
+
+    /**
+     * Retrieves a subscription by its name without checking if it is active.
+     *
+     * @param subscriptionName the name of the subscription
+     *
+     * @return the subscription
+     */
+    public VaradhiSubscription getSubscriptionWithoutValidation(String subscriptionName) {
```
Comment (gauravAshok) [inline server/src/main/java/com/flipkart/varadhi/services/SubscriptionService.java:80]: i think you can keep the withoutValidation as standard. Dhruv commented the same that getSubscription() method probably wont require validation.

File: server/src/main/java/com/flipkart/varadhi/services/SubscriptionService.java [hunk 12]
```diff
@@ -1,46 +1,107 @@
...
-        validateCreation(subscribedTopic, subscription);
+        validateGroupedSubscription(subscribedTopic, subscription);
         metaStore.createSubscription(subscription);
+
         try {
             shardProvisioner.provision(subscription, subProject);
             subscription.markCreated();
         } catch (Exception e) {
-            log.error("create subscription failed", e);
+            log.error("Failed to create subscription", e);
```
Comment (gauravAshok) [inline server/src/main/java/com/flipkart/varadhi/services/SubscriptionService.java:104]: logging and throwing together is redundant. 
dont log if you are throwing.

File: server/src/main/java/com/flipkart/varadhi/services/SubscriptionService.java (hunk 11 above)
Comment (gauravAshok) [reply server/src/main/java/com/flipkart/varadhi/services/SubscriptionService.java:80]: The `active` validation can be done on the result by the caller based on its use case.

File: server/src/main/java/com/flipkart/varadhi/services/SubscriptionService.java [hunk 13]
```diff
@@ -49,89 +110,270 @@ public VaradhiSubscription createSubscription(
...
-        validateForConflictingUpdate(fromVersion, existingSubscription.getVersion());
-        VaradhiTopic subscribedTopic = metaStore.getTopic(existingSubscription.getTopic());
-        validateForSubscribedTopic(subscribedTopic, grouped);
-        return controllerClient.getSubscriptionState(subscriptionName, requestedBy).thenApply(ss -> {
-            existingSubscription.setGrouped(grouped);
-            existingSubscription.setDescription(description);
-            existingSubscription.setEndpoint(endpoint);
-            existingSubscription.setRetryPolicy(retryPolicy);
-            existingSubscription.setConsumptionPolicy(consumptionPolicy);
+        VaradhiSubscription subscription = getValidatedSubscription(subscriptionName);
```
Comment (gauravAshok) [inline server/src/main/java/com/flipkart/varadhi/services/SubscriptionService.java:154]: is validatedSubscription a requirement for updates?

File: server/src/main/java/com/flipkart/varadhi/services/SubscriptionService.java [hunk 14]
```diff
@@ -49,89 +110,270 @@ public VaradhiSubscription createSubscription(
...
+        validateVersionForUpdate(fromVersion, subscription.getVersion());
+
+        boolean originalGrouped = subscription.isGrouped();
+
+        subscription.setGrouped(grouped);
+        validateGroupedSubscription(metaStore.getTopic(subscription.getTopic()), subscription);
+
+        subscription.setGrouped(originalGrouped);
```
Comment (gauravAshok) [inline server/src/main/java/com/flipkart/varadhi/services/SubscriptionService.java:162]: why are we setting originalGrouped again?

Comment (gauravAshok) [reply server/src/main/java/com/flipkart/varadhi/services/SubscriptionService.java:162]: Modifying the returned value from the method is risky. 
We should clone it before modifying.

Lets follow the receiver of the object is supposed to make a copy.
For eg: If the method wants to modify the passed parameters, then the method (callee) is supposed to make the copy. 
If the caller wants to modify the returned value from the method, then the caller needs to make the copy.

File: server/src/main/java/com/flipkart/varadhi/services/VaradhiTopicService.java [hunk 15]
```diff
@@ -1,74 +1,208 @@
...
+     * @return the Varadhi topic
+     * @throws ResourceNotFoundException if the topic is not found or inactive
+     */
     public VaradhiTopic get(String topicName) {
-        return metaStore.getTopic(topicName);
+        VaradhiTopic varadhiTopic = metaStore.getTopic(topicName);
+
+        if (!varadhiTopic.isActive()) {
+            throw new ResourceNotFoundException("Topic %s not found.".formatted(topicName));
+        }
```
Comment (gauravAshok) [inline server/src/main/java/com/flipkart/varadhi/services/VaradhiTopicService.java:67]: lets remove this check. lets keep the normal get call unguarded.

File: server/src/main/java/com/flipkart/varadhi/web/entities/ResourceActionRequest.java [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+ * @param actorCode the action code indicating the action to be performed, must not be null
+ * @param message    an optional message associated with the action
+ */
+public record ResourceActionRequest(@NotNull LifecycleStatus.ActorCode actorCode, String message)
+        implements Validatable {
+
+    public ResourceActionRequest {
+        if (actorCode == null) {
+            throw new NullPointerException("actionCode must not be null");
+        }
```
Comment (gauravAshok) [inline server/src/main/java/com/flipkart/varadhi/web/entities/ResourceActionRequest.java:19]: object.requiresNonNull

File: server/src/main/java/com/flipkart/varadhi/web/v1/admin/SubscriptionHandlers.java [hunk 17]
```diff
@@ -20,49 +26,77 @@
...
-import static com.flipkart.varadhi.Constants.PathParams.*;
-import static com.flipkart.varadhi.entities.Hierarchies.*;
+import static com.flipkart.varadhi.Constants.PathParams.PATH_PARAM_PROJECT;
+import static com.flipkart.varadhi.Constants.PathParams.PATH_PARAM_SUBSCRIPTION;
+import static com.flipkart.varadhi.Constants.QueryParams.QUERY_PARAM_DELETION_TYPE;
+import static com.flipkart.varadhi.Constants.QueryParams.QUERY_PARAM_IGNORE_CONSTRAINTS;
+import static com.flipkart.varadhi.Constants.QueryParams.QUERY_PARAM_INCLUDE_INACTIVE;
+import static com.flipkart.varadhi.Constants.QueryParams.QUERY_PARAM_MESSAGE;
+import static com.flipkart.varadhi.entities.Hierarchies.SubscriptionHierarchy;
+import static com.flipkart.varadhi.entities.Hierarchies.TopicHierarchy;
```
Comment (gauravAshok) [inline server/src/main/java/com/flipkart/varadhi/web/v1/admin/SubscriptionHandlers.java:41]: import static *?

File: server/src/main/java/com/flipkart/varadhi/web/v1/admin/SubscriptionHandlers.java [hunk 18]
```diff
@@ -20,49 +26,77 @@
...
 import static com.flipkart.varadhi.entities.VersionedEntity.NAME_SEPARATOR;
 import static com.flipkart.varadhi.entities.VersionedEntity.NAME_SEPARATOR_REGEX;
-import static com.flipkart.varadhi.entities.auth.ResourceAction.*;
+import static com.flipkart.varadhi.entities.auth.ResourceAction.SUBSCRIPTION_CREATE;
+import static com.flipkart.varadhi.entities.auth.ResourceAction.SUBSCRIPTION_DELETE;
+import static com.flipkart.varadhi.entities.auth.ResourceAction.SUBSCRIPTION_GET;
+import static com.flipkart.varadhi.entities.auth.ResourceAction.SUBSCRIPTION_LIST;
```
Comment (gauravAshok) [inline server/src/main/java/com/flipkart/varadhi/web/v1/admin/SubscriptionHandlers.java:47]: import static *?

File: server/src/main/java/com/flipkart/varadhi/web/v1/admin/SubscriptionHandlers.java [hunk 19]
```diff
@@ -20,49 +26,77 @@
...
+import static com.flipkart.varadhi.entities.auth.ResourceAction.SUBSCRIPTION_RESTORE;
```
Comment (gauravAshok) [inline server/src/main/java/com/flipkart/varadhi/web/v1/admin/SubscriptionHandlers.java:48]: Why is restore an action?
I think we can use SUBSCRIPTION_UPDATE auth here.

File: server/src/main/java/com/flipkart/varadhi/web/v1/admin/TopicHandlers.java [hunk 20]
```diff
@@ -17,25 +21,44 @@
...
+import static com.flipkart.varadhi.entities.Hierarchies.ProjectHierarchy;
+import static com.flipkart.varadhi.entities.Hierarchies.TopicHierarchy;
 import static com.flipkart.varadhi.entities.VersionedEntity.NAME_SEPARATOR;
 import static com.flipkart.varadhi.entities.VersionedEntity.NAME_SEPARATOR_REGEX;
-import static com.flipkart.varadhi.entities.auth.ResourceAction.*;
+import static com.flipkart.varadhi.entities.auth.ResourceAction.TOPIC_CREATE;
+import static com.flipkart.varadhi.entities.auth.ResourceAction.TOPIC_DELETE;
+import static com.flipkart.varadhi.entities.auth.ResourceAction.TOPIC_GET;
+import static com.flipkart.varadhi.entities.auth.ResourceAction.TOPIC_LIST;
+import static com.flipkart.varadhi.entities.auth.ResourceAction.TOPIC_RESTORE;
```
Comment (gauravAshok) [inline server/src/main/java/com/flipkart/varadhi/web/v1/admin/TopicHandlers.java:42]: same here. TOPIC_UPDATE should be enough

**Review by gauravAshok — COMMENTED** [review]
reviewed.

File: server/src/main/java/com/flipkart/varadhi/services/SubscriptionService.java (hunk 9 above)
Comment (devanshu-garg-IN) [reply server/src/main/java/com/flipkart/varadhi/services/SubscriptionService.java:258]: Agreed! Done.

File: server/src/main/java/com/flipkart/varadhi/services/SubscriptionService.java (hunk 10 above)
Comment (devanshu-garg-IN) [reply server/src/main/java/com/flipkart/varadhi/services/SubscriptionService.java:57]: Done.

File: server/src/main/java/com/flipkart/varadhi/services/SubscriptionService.java (hunk 11 above)
Comment (devanshu-garg-IN) [reply server/src/main/java/com/flipkart/varadhi/services/SubscriptionService.java:80]: Understood, done.

File: server/src/main/java/com/flipkart/varadhi/services/SubscriptionService.java (hunk 13 above)
Comment (devanshu-garg-IN) [reply server/src/main/java/com/flipkart/varadhi/services/SubscriptionService.java:154]: Yeah, ideally, if a subscription is in an errored or inactive state, updates should either be restricted or have no meaningful impact.

File: server/src/main/java/com/flipkart/varadhi/services/VaradhiTopicService.java (hunk 15 above)
Comment (devanshu-garg-IN) [reply server/src/main/java/com/flipkart/varadhi/services/VaradhiTopicService.java:67]: Done

File: server/src/main/java/com/flipkart/varadhi/web/entities/ResourceActionRequest.java (hunk 16 above)
Comment (devanshu-garg-IN) [reply server/src/main/java/com/flipkart/varadhi/web/entities/ResourceActionRequest.java:19]: Done

File: server/src/main/java/com/flipkart/varadhi/web/v1/admin/TopicHandlers.java (hunk 20 above)
Comment (devanshu-garg-IN) [reply server/src/main/java/com/flipkart/varadhi/web/v1/admin/TopicHandlers.java:42]: Done

File: server/src/main/java/com/flipkart/varadhi/web/v1/admin/SubscriptionHandlers.java (hunk 19 above)
Comment (devanshu-garg-IN) [reply server/src/main/java/com/flipkart/varadhi/web/v1/admin/SubscriptionHandlers.java:48]: Done

File: server/src/main/java/com/flipkart/varadhi/web/v1/admin/SubscriptionHandlers.java (hunk 18 above)
Comment (devanshu-garg-IN) [reply server/src/main/java/com/flipkart/varadhi/web/v1/admin/SubscriptionHandlers.java:47]: Wildcard imports are forbidden by §3.3.1 of the Google Java Style Guide. Although, I can bypass it.
